### PR TITLE
76961 govuk pay client

### PR DESCRIPTION
**Pull request by JohnPadgettCivica**
The bulk of the happy path GOV.UK integration. Need  finish up unit and integration tests.

### CODE REVIEW

**Comment by JohnPadgettCivica** [comment]
Need to add unit and integration tests for this.

File: src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentFailure.cshtml [hunk 1]
```diff
@@ -0,0 +1 @@
+﻿
```
Comment (will-hope98) [inline src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentFailure.cshtml:1]: Should this at least containt some title for the sake of testers confirming the redirect was correct

File: src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentSuccess.cshtml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿@using System.Web.Mvc.Html
+@using EA.Weee.Web.Controllers
+@using EA.Weee.Web.RazorHelpers
+@using MvcNavigationHelpers
+@model EA.Weee.Web.Areas.Producer.ViewModels.PaymentResultModel
+@{
+    var title = "You have successfully submitted your producer registration";
```
Comment (will-hope98) [inline src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentSuccess.cshtml:7]: Can't find a wireframe for this view anywhere

File: src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentResult.cshtml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+﻿
```
Comment (will-hope98) [inline src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentResult.cshtml:1]: What's this view for?  Isn't this the screen Steve will have been working on

File: src/EA.Weee.Web/EA.Weee.Web.csproj [hunk 4]
```diff
@@ -2144,10 +2156,6 @@
       <Project>{A987FFF6-5386-419E-AF0A-E6C74E3323CF}</Project>
       <Name>EA.Weee.Domain</Name>
     </ProjectReference>
-    <ProjectReference Include="..\EA.Weee.RequestHandlers\EA.Weee.RequestHandlers.csproj">
```
Comment (will-hope98) [inline src/EA.Weee.Web/EA.Weee.Web.csproj:2147]: Should this be removed?  Is it not used by other classes in here

File: src/EA.Weee.Requests/Organisations/DirectRegistrant/UpdateSubmissionPaymentDetailsRequest.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    public class UpdateSubmissionPaymentDetailsRequest : IRequest<bool>
+    {
+        public Guid DirectRegistrantId { get; private set; }
+
+        public PaymentStatus PaymentStatus { get; private set; }
+
+        public Guid PaymentSessionId { get; private set; }
+
+        public bool IsFinalState { get; private set; }
```
Comment (will-hope98) [inline src/EA.Weee.Requests/Organisations/DirectRegistrant/UpdateSubmissionPaymentDetailsRequest.cs:16]: Should these not be nullable if they're not mandatory, and have no Condition.Requires()

File: src/EA.Weee.Domain/Producer/DirectProducerSubmission.cs [hunk 6]
```diff
@@ -33,13 +29,22 @@ public DirectProducerSubmission()
 
         public virtual ICollection<DirectProducerSubmissionHistory> SubmissionHistory { get; set; }
 
+        public virtual PaymentSession FinalPaymentSession { get; set; }
+
+        //public virtual DateTime PaymentStartedDateTime { get; set; }
```
Comment (will-hope98) [inline src/EA.Weee.Domain/Producer/DirectProducerSubmission.cs:34]: Can this be removed?

File: src/EA.Weee.DataAccess/Mappings/DirectProducerSubmissionMapping.cs [hunk 7]
```diff
@@ -9,10 +9,10 @@ public DirectProducerSubmissionConfiguration()
         {
             ToTable("DirectProducerSubmission", "Producer");
 
-            Property(e => e.PaymentStatus).IsOptional();
-            Property(e => e.PaymentReference).HasMaxLength(20).IsOptional();
-            Property(e => e.PaymentId).HasMaxLength(20).IsOptional();
+            Property(e => e.PaymentFinished).IsOptional();
+            
             Property(e => e.ComplianceYear).IsRequired();
+            //Property(e => e.PaymentStartedDateTime).IsOptional();
```
Comment (will-hope98) [inline src/EA.Weee.DataAccess/Mappings/DirectProducerSubmissionMapping.cs:15]: Should this be removed if unused

File: src/EA.Weee.Api.Client/Models/Pay/PaymentWithAllLinks.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        [JsonPropertyName("created_date")]
+        public DateTime CreatedDate { get; set; }
+
+        [JsonPropertyName("refund_summary")]
+        public RefundSummary RefundSummary { get; set; }
+
+        [JsonPropertyName("settlement_summary")]
+        public SettlementSummary SettlementSummary { get; set; }
+
+        [JsonPropertyName("_links")]
```
Comment (will-hope98) [inline src/EA.Weee.Api.Client/Models/Pay/PaymentWithAllLinks.cs:38]: Should there be an _ here?

File: src/EA.Weee.Api.Client/Models/Pay/CreatePaymentResult.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        [JsonPropertyName("return_url")]
+        public string ReturnUrl { get; set; }
+
+        [JsonPropertyName("reference")]
+        public string Reference { get; set; }
+
+        [JsonPropertyName("created_date")]
+        public DateTime CreatedDate { get; set; }
+
+        [JsonPropertyName("_links")]
```
Comment (will-hope98) [inline src/EA.Weee.Api.Client/Models/Pay/CreatePaymentResult.cs:29]: Should therebe an _ here?

**Comment by will-hope98** [comment]
This branch also has an error when clicking new/continue registration on an existing organisation.  It throws an api exception.

File: src/EA.Weee.Web/EA.Weee.Web.csproj (hunk 4 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Web/EA.Weee.Web.csproj:2147]: It looks to have been added accidently t some point.

File: src/EA.Weee.Api.Client/Models/Pay/CreatePaymentResult.cs (hunk 9 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Api.Client/Models/Pay/CreatePaymentResult.cs:29]: Yeah this is their API naming...

File: src/EA.Weee.Api.Client/Models/Pay/PaymentWithAllLinks.cs (hunk 8 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Api.Client/Models/Pay/PaymentWithAllLinks.cs:38]: Yeah this is their (GOV.UK pay) API naming...

File: src/EA.Weee.Requests/Organisations/DirectRegistrant/UpdateSubmissionPaymentDetailsRequest.cs (hunk 5 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Requests/Organisations/DirectRegistrant/UpdateSubmissionPaymentDetailsRequest.cs:16]: The conditions check that the Guids are not the default value. The enum won't be null and the isFinalState should always be a true or false.

File: src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentResult.cshtml (hunk 3 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentResult.cshtml:1]: Removed.

File: src/EA.Weee.Domain/Producer/DirectProducerSubmission.cs (hunk 6 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Domain/Producer/DirectProducerSubmission.cs]: removed

File: src/EA.Weee.DataAccess/Mappings/DirectProducerSubmissionMapping.cs (hunk 7 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.DataAccess/Mappings/DirectProducerSubmissionMapping.cs]: removed.

File: src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentFailure.cshtml (hunk 1 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentFailure.cshtml]: updated.

**Comment by JohnPadgettCivica** [comment]
> This branch also has an error when clicking new/continue registration on an existing organisation. It throws an api exception.

Suspect this was failing as there are some DB updates... but dont worry about those until this is in develop. Which could be worth getting merged when you are happy with comment resolution.

File: src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentSuccess.cshtml (hunk 2 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Web/Areas/Producer/Views/ProducerSubmission/PaymentSuccess.cshtml:7]: https://liysak.axshare.com/?id=5f29oz&p=76379_-_confirmation_page__1_&g=1

**Comment by will-hope98** [comment]
> > This branch also has an error when clicking new/continue registration on an existing organisation. It throws an api exception.
> 
> Suspect this was failing as there are some DB updates... but dont worry about those until this is in develop. Which could be worth getting merged when you are happy with comment resolution.

Yh looks good now

**Pull request merged**